rfcs/block-component.md

```diff
@@ -0,0 +1,78 @@
+# Block Component
+
+One of the most common needs when building web apps is creating layout elements that have certain layout and typography styles. 
+
+For more traditional apps that use global stylesheets, many of the popular CSS frameworks have helper classes to quickly create styled elements (see [Bootstrap](https://getbootstrap.com/docs/4.1/utilities/sizing/), [Tachyons](http://tachyons.io/docs/layout/spacing/)). Uber's superfine css framework also had this and it was used very widely–just search something like `push-large` in sourcegraph.
+
+The move to CSS-in-JS sort of eliminates these classes in favor of encapsulating styles within components. This is a generally a good thing because it promotes reusability, however the reality is that there are dozens or hundreds of different layout/spacing/typography combinations in any app, and there's not always a clear semantic meaning or potential for reuse. In practice you end up just defining and naming a bunch of things at the top of the file:
+
+```js
+const PricingPageHero = styled('div', {
+  margin: '10px 15px',
+  color: '#ccc',
+  display: 'flex',
+});
+
+const PricingPageHeroText = styled('h2', {
+  margin: '10px 0',
+});
+
+// ... and many more
+```
+
+---
+
+## Proposal
+
+### `<Block/>` Component
+
+The `Block` component is a helper component that lets you write these common less-reusable elements more succinctly inline.
+
+The above example might turn into:
+
+```js
+<Block marginSides="scale200" marginEnds="scale150" color="grey200" display="flex" flexDirection="column">
+  <Block as="h2" margin="scale150" justifySelf="left">
+    Simple Pricing
+  </Block>
+  {/* ... etc */}
+</Block>
+```
+
+Props would be exposed for:
+- Changing base element
+- Spacing (margin/padding)
```

Comment:
Thoughts: margin should be separated into marginTop, marginRight, marginLeft, and marginBottom, whereas padding should only have symmetric (padding, paddingX, and paddingY) attributes, to emphasize the differences between the two properties and proper usage.

Reply:
Padding doesn't have to be symmetric, it's not in the rules. It's more about making sure there's only one way to do things. If you want an asymmetric div in your box model, use margins (esp. since they can be negative), otherwise use padding. It maintains a "consistent visual rhythm" - from http://basscss.com/v7/docs/white-space/

Reply:
I feel like good practice should be documented and not enforced via code. We can give guidance on what's the best approach to using the component and doing layout, but if people really want to roll with their own ideas they should be able to do it.